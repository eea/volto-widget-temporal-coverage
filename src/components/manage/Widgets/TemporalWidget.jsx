import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { Grid } from 'semantic-ui-react';
import { defineMessages, injectIntl } from 'react-intl';
import CreatableSelect from 'react-select/creatable';
import { FormFieldWrapper } from '@plone/volto/components';
import {
  Option,
  DropdownIndicator,
  selectTheme,
  customSelectStyles,
} from '@plone/volto/components/manage/Widgets/SelectStyling';

import {
  createTemporalRangeOptions,
  getIndividualValues,
  createOption,
} from '@eeacms/volto-widget-temporal-coverage/helpers';

import { TemporalWidgetView } from '@eeacms/volto-widget-temporal-coverage/components';

import './public.less';

const messages = defineMessages({
  temporalCoverage: {
    id: 'Temporal coverage',
    defaultMessage: 'Temporal coverage',
  },
  NoSelection: {
    id: 'No selection',
    defaultMessage: 'No selection',
  },
});

function addTemporalValues(current_temporal_values, new_values) {
  let year_values = current_temporal_values
    ? current_temporal_values.map((item) => item.value)
    : [];
  if (new_values.includes('-')) {
    let split_values = new_values.split('-');
    let start_year = parseInt(split_values[0]);
    let end_year = parseInt(split_values[1]);
    for (let year = start_year; year <= end_year; year++) {
      if (year_values.indexOf(year) === -1) {
        year_values.push(year);
      }
    }
  } else {
    const year = parseInt(new_values);
    if (year_values.indexOf(year) === -1) {
      year_values.push(year);
    }
  }
  year_values.sort();
  return year_values.map((year) => createOption(year));
}

const TemporalWidget = (props) => {
  const {
    onChange,
    value,
    intl,
    id = 'select-temporal-coverage',
    title = 'Temporal coverage',
  } = props;

  const [currentInputValue, setCurrentInputValue] = React.useState('');
  const [temporalRangeOptions, setTemporalRangeOptions] = React.useState([]);

  useEffect(() => {
    setTemporalRangeOptions(
      createTemporalRangeOptions(getIndividualValues(value?.temporal)),
    );
  }, [value]);

  return (
    <FormFieldWrapper
      {...props}
      id={id}
      className="temporal-field-wrapper"
      title={title}
      columns={1}
    >
      <Grid>
        <Grid.Row stretched>
          <Grid.Column width="4">
            <div className="wrapper">
              <label htmlFor="select-listingblock-template">
                {intl.formatMessage(messages.temporalCoverage)}
              </label>
            </div>
          </Grid.Column>
          <Grid.Column width="8" style={{ flexDirection: 'unset' }}>
            {value?.readOnly ? (
              <TemporalWidgetView value={value} className="read-only" />
            ) : (
              <CreatableSelect
                defaultValue={temporalRangeOptions}
                isMulti
                allowCreateWhileLoading={true}
                id={id}
                inputValue={currentInputValue}
                name="select-temporal-coverage"
                className="react-select-container"
                classNamePrefix="react-select"
                options={[
                  {
                    label: intl.formatMessage(messages.NoSelection),
                    value: '',
                  },
                ]}
                onInputChange={(newInputValue) => {
                  if (!newInputValue) {
                    return setCurrentInputValue(newInputValue);
                  }
                  let new_input_length = newInputValue.length;
                  let last_char = newInputValue[new_input_length - 1];
                  // dissallow non numeric values and allow - only as 5th char
                  if (
                    (last_char === '-' && new_input_length !== 5) ||
                    (['-', '0'].indexOf(last_char) === -1 &&
                      !parseInt(last_char))
                  ) {
                    return currentInputValue;
                  }

                  if (
                    new_input_length === 5 &&
                    new_input_length > currentInputValue.length &&
                    newInputValue[4] !== '-'
                  ) {
                    newInputValue =
                      newInputValue.slice(0, 4) +
                      '-' +
                      newInputValue.slice(4, 5);
                  }
                  if (new_input_length <= 9) {
                    return setCurrentInputValue(newInputValue);
                  }

                  return setCurrentInputValue(newInputValue);
                }}
                isValidNewOption={(inputValue) => {
                  let new_option = inputValue.split('-');
                  // allow only ranges when second value is higher
                  if (new_option.length === 2) {
                    let first_value = parseInt(new_option[0]);
                    let second_value = parseInt(new_option[1]);
                    return first_value < second_value;
                  }
                  return /^\d+$/.test(parseInt(inputValue.split('-')[0]));
                }}
                onCreateOption={(newOption) => {
                  let temporal_values = addTemporalValues(
                    getIndividualValues(value?.temporal),
                    newOption,
                  );
                  onChange(
                    id,
                    newOption === ''
                      ? undefined
                      : { temporal: temporal_values },
                  );
                }}
                value={temporalRangeOptions}
                styles={customSelectStyles}
                theme={selectTheme}
                components={{ DropdownIndicator, Option }}
                onChange={(values) => {
                  let temporal_values =
                    (value.length && getIndividualValues(values)) || values;
                  onChange(
                    id,
                    value === '' ? undefined : { temporal: temporal_values },
                  );
                }}
              />
            )}
          </Grid.Column>
        </Grid.Row>
      </Grid>
    </FormFieldWrapper>
  );
};

TemporalWidget.propTypes = {
  value: PropTypes.objectOf(PropTypes.any),
  block: PropTypes.string,
};

export default injectIntl(TemporalWidget);
