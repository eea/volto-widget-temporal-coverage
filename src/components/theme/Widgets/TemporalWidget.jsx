import React from 'react';
import cx from 'classnames';
import { createTemporalRangeOptions } from '@eeacms/volto-widget-temporal-coverage/helpers';

import './public.less';

const TemporalWidget = ({ value, children, className }) => {
  const temporal_ranges = createTemporalRangeOptions(value);
  return temporal_ranges.length ? (
    <div className={cx(className, 'temporal-coverage', 'widget')}>
      {temporal_ranges.map((item, index) => (
        <div key={index} className={'temporal-coverage-value'}>
          {children && children instanceof Function
            ? children(item?.label)
            : item?.label}
        </div>
      ))}
    </div>
  ) : null;
};
export default TemporalWidget;
