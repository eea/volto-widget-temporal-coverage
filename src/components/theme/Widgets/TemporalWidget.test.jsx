import React from 'react';
import renderer from 'react-test-renderer';
import TemporalWidget from './TemporalWidget';

describe('TemporalWidget view tests', () => {
  it('renders an empty widget when value is an empty list', () => {
    const component = renderer.create(<TemporalWidget value={[]} />);
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('renders an empty widget when value is null', () => {
    const component = renderer.create(<TemporalWidget value={null} />);
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('renders a widget when passing a single temporal value', () => {
    const component = renderer.create(
      <TemporalWidget value={[{ value: '1900', label: '1900' }]} />,
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('renders a widget with multiple temporal values', () => {
    const component = renderer.create(
      <TemporalWidget
        value={[
          { value: '1900', label: '1900' },
          { value: '1902', label: '1902' },
        ]}
      />,
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('renders a widget with temporal ranges', () => {
    const component = renderer.create(
      <TemporalWidget
        value={[
          { value: '1900', label: '1900' },
          { value: '1901', label: '1901' },
        ]}
      />,
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('renders a widget with multiple temporal ranges', () => {
    const component = renderer.create(
      <TemporalWidget
        value={[
          { value: '1900', label: '1900' },
          { value: '1901', label: '1901' },
          { value: '1903', label: '1903' },
          { value: '1904', label: '1904' },
        ]}
      />,
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('renders a widget with temporal ranges and single values', () => {
    const component = renderer.create(
      <TemporalWidget
        value={[
          { value: '1900', label: '1900' },
          { value: '1901', label: '1901' },
          { value: '1902', label: '1902' },
          { value: '1910', label: '1910' },
        ]}
      />,
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('renders a widget with existing temporal ranges and single values', () => {
    const component = renderer.create(
      <TemporalWidget
        value={[
          { value: '1900-1904', label: '1900-1904' },
          { value: '1910', label: '1910' },
        ]}
      />,
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('renders a widget with multiple single values and  multiple temporal ranges', () => {
    const component = renderer.create(
      <TemporalWidget
        value={[
          { value: '1890', label: '1890' },
          { value: '1900-1905', label: '1900-1905' },
          { value: '1915', label: '1915' },
          { value: '1920-1922', label: '1920-1922' },
        ]}
      />,
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('renders a widget with multiple single values and temporal ranges where one single value is removed due to the previous range value', () => {
    const component = renderer.create(
      <TemporalWidget
        value={[
          { value: '1900-1905', label: '1900-1905' },
          { value: '1904', label: '1904' },
          { value: '1907', label: '1907' },
        ]}
      />,
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('renders a widget with a custom css class', () => {
    const component = renderer.create(
      <TemporalWidget
        className="custom-css-class"
        value={[{ value: '1900', label: '1900' }]}
      />,
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('renders a widget where the values are wrapped by custom child tags', () => {
    const component = renderer.create(
      <TemporalWidget
        value={[
          { value: '1890', label: '1890' },
          { value: '1900-1905', label: '1900-1905' },
        ]}
      >
        {(child) => <strong>{child}</strong>}
      </TemporalWidget>,
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });
});
