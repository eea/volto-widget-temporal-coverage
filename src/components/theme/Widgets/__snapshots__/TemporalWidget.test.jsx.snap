// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TemporalWidget view tests renders a widget when passing a single temporal value 1`] = `
<div
  className="temporal-coverage widget"
>
  <div
    className="temporal-coverage-value"
  >
    1900
  </div>
</div>
`;

exports[`TemporalWidget view tests renders a widget where the values are wrapped by custom child tags 1`] = `
<div
  className="temporal-coverage widget"
>
  <div
    className="temporal-coverage-value"
  >
    <strong>
      1890
    </strong>
  </div>
  <div
    className="temporal-coverage-value"
  >
    <strong>
      1900-1905
    </strong>
  </div>
</div>
`;

exports[`TemporalWidget view tests renders a widget with a custom css class 1`] = `
<div
  className="custom-css-class temporal-coverage widget"
>
  <div
    className="temporal-coverage-value"
  >
    1900
  </div>
</div>
`;

exports[`TemporalWidget view tests renders a widget with existing temporal ranges and single values 1`] = `
<div
  className="temporal-coverage widget"
>
  <div
    className="temporal-coverage-value"
  >
    1900-1904
  </div>
  <div
    className="temporal-coverage-value"
  >
    1910
  </div>
</div>
`;

exports[`TemporalWidget view tests renders a widget with multiple single values and  multiple temporal ranges 1`] = `
<div
  className="temporal-coverage widget"
>
  <div
    className="temporal-coverage-value"
  >
    1890
  </div>
  <div
    className="temporal-coverage-value"
  >
    1900-1905
  </div>
  <div
    className="temporal-coverage-value"
  >
    1915
  </div>
  <div
    className="temporal-coverage-value"
  >
    1920-1922
  </div>
</div>
`;

exports[`TemporalWidget view tests renders a widget with multiple single values and temporal ranges where one single value is removed due to the previous range value 1`] = `
<div
  className="temporal-coverage widget"
>
  <div
    className="temporal-coverage-value"
  >
    1900-1905
  </div>
  <div
    className="temporal-coverage-value"
  >
    1907
  </div>
</div>
`;

exports[`TemporalWidget view tests renders a widget with multiple temporal ranges 1`] = `
<div
  className="temporal-coverage widget"
>
  <div
    className="temporal-coverage-value"
  >
    1900-1901
  </div>
  <div
    className="temporal-coverage-value"
  >
    1903-1904
  </div>
</div>
`;

exports[`TemporalWidget view tests renders a widget with multiple temporal values 1`] = `
<div
  className="temporal-coverage widget"
>
  <div
    className="temporal-coverage-value"
  >
    1900
  </div>
  <div
    className="temporal-coverage-value"
  >
    1902
  </div>
</div>
`;

exports[`TemporalWidget view tests renders a widget with temporal ranges 1`] = `
<div
  className="temporal-coverage widget"
>
  <div
    className="temporal-coverage-value"
  >
    1900-1901
  </div>
</div>
`;

exports[`TemporalWidget view tests renders a widget with temporal ranges and single values 1`] = `
<div
  className="temporal-coverage widget"
>
  <div
    className="temporal-coverage-value"
  >
    1900-1902
  </div>
  <div
    className="temporal-coverage-value"
  >
    1910
  </div>
</div>
`;

exports[`TemporalWidget view tests renders an empty widget when value is an empty list 1`] = `null`;

exports[`TemporalWidget view tests renders an empty widget when value is null 1`] = `null`;
